picoquic/packet.c

```diff
@@ -40,199 +41,167 @@ int picoquic_parse_long_packet_header(
     size_t length,
     struct sockaddr* addr_from,
     picoquic_packet_header* ph,
-    picoquic_cnx_t** pcnx,
-    int receiving)
+    picoquic_cnx_t** pcnx)
 {
     int ret = 0;
+    bytestream stream;
+    bytestream * s = bytestream_ref_init(&stream, bytes, length);
 
-    if (length < 6) {
-        ret = -1;
-    }
-    else {
-        /* The bytes at position 1..4 describe the version */
-        ph->vn = PICOPARSE_32(bytes + 1);
-        uint8_t l_dest_id = 0;
-        uint8_t l_srce_id = 0;
-        uint32_t i_srce_id = 0;
+    uint8_t flags = 0;
+    ret |= byteread_int8(s, &flags);
+    ret |= byteread_int32(s, &ph->vn);
+    ret |= byteread_cid(s, &ph->dest_cnx_id);
+    ret |= byteread_cid(s, &ph->srce_cnx_id);
 
```

Comment:
Compare that to using the constructs used in frame.c, such as picoquic_frames_uint8_decode, etc. You would write:

```
uint8_t * bytes_max = bytes + length;

if ((bytes = picoquic_frames_uint8_decode(bytes, bytes_max, &flags)) == NULL ||
    (bytes = picoquic_frames_uint32_decode(bytes, bytes_max, &ph->vn)) == NULL  ||
    (bytes = picoquic_frames_cid_decode(bytes, bytes_max, &ph->dest_cnx_id)) == NULL ||
    (bytes = picoquic_frames_cid_decode(bytes, bytes_max, &ph->srce_cnx_id)) == NULL){
          ret = some_error_code;
}
```

Avoiding the use of a specific stream construct, and also stopping parsing as soon as an error is detected.

On the other hand, I don't want to discourage progress, I know that the differences are minimal, and I also know that we are not anywhere near the critical path.

I guess my real beef is with the `ret |= ` construct, which is not right because ret is not a boolean, and also does not stop parsing as soon as an error is detected.

```
if ((ret = byteread_int8(s, &flags)) == 0 &&
    (ret |= byteread_int32(s, &ph->vn)) == 0 &&
    (ret |= byteread_cid(s, &ph->dest_cnx_id)) == 0 &&
    (ret |= byteread_cid(s, &ph->srce_cnx_id)) {
          /* continue with processing */
```
would work.

Reply:
Thanks, Christian, for your feedback. I totally agree with your comment about the `ret |= ...` construct and I'll change that.

The other thing is just a matter of taste and I prefer to implement your taste as you're the maintainer of picoquic :) So, I'll change that too.

Reply:
Thanks for being flexible. Apart from taste, there is also the issue of rewriting frame.c or not. If we pick a standard way, we will have to apply it everywhere.

The _frames_ functions were suggested by @igorlord, as a way to shorten the size of the frame decoding functions. However, we did it halfway, and only considered the decoding functions. I suppose I have now to come up with the "picoquic_frames_*_encode()" functions.

Reply:
Fixing frames.c decoding is on my list already. I also will see if I can change the bytestream stuff to the picoquic_*_decode functions.